pkg/cloudprovider/routecontroller/routecontroller_test.go

```diff
@@ -0,0 +1,200 @@
+			},
+			initialRoutes: []*cloudprovider.Route{
+				{cluster + "-01", "node-1", "10.120.0.0/24", "k8s-node-route"},
+			},
+			expectedRoutes: []*cloudprovider.Route{
+				{cluster + "-01", "node-1", "10.120.0.0/24", "k8s-node-route"},
+				{cluster + "-02", "node-2", "10.120.1.0/24", "k8s-node-route"},
+			},
+		},
+		// 2 nodes, no routes yet
+		{
+			nodes: []api.Node{
+				{ObjectMeta: api.ObjectMeta{Name: "node-1", UID: "01"}, Spec: api.NodeSpec{PodCIDR: "10.120.0.0/24"}},
+				{ObjectMeta: api.ObjectMeta{Name: "node-2", UID: "02"}, Spec: api.NodeSpec{PodCIDR: "10.120.1.0/24"}},
+			},
+			initialRoutes: []*cloudprovider.Route{},
+			expectedRoutes: []*cloudprovider.Route{
+				{cluster + "-01", "node-1", "10.120.0.0/24", "k8s-node-route"},
+				{cluster + "-02", "node-2", "10.120.1.0/24", "k8s-node-route"},
+			},
+		},
+		// 2 nodes, a few too many routes
+		{
+			nodes: []api.Node{
+				{ObjectMeta: api.ObjectMeta{Name: "node-1", UID: "01"}, Spec: api.NodeSpec{PodCIDR: "10.120.0.0/24"}},
+				{ObjectMeta: api.ObjectMeta{Name: "node-2", UID: "02"}, Spec: api.NodeSpec{PodCIDR: "10.120.1.0/24"}},
+			},
+			initialRoutes: []*cloudprovider.Route{
+				{cluster + "-01", "node-1", "10.120.0.0/24", "k8s-node-route"},
+				{cluster + "-02", "node-2", "10.120.1.0/24", "k8s-node-route"},
+				{cluster + "-03", "node-3", "10.120.2.0/24", "k8s-node-route"},
+				{cluster + "-04", "node-4", "10.120.3.0/24", "k8s-node-route"},
+			},
+			expectedRoutes: []*cloudprovider.Route{
+				{cluster + "-01", "node-1", "10.120.0.0/24", "k8s-node-route"},
+				{cluster + "-02", "node-2", "10.120.1.0/24", "k8s-node-route"},
+			},
+		},
+		// 2 nodes, 2 routes, but only 1 is right
+		{
+			nodes: []api.Node{
+				{ObjectMeta: api.ObjectMeta{Name: "node-1", UID: "01"}, Spec: api.NodeSpec{PodCIDR: "10.120.0.0/24"}},
+				{ObjectMeta: api.ObjectMeta{Name: "node-2", UID: "02"}, Spec: api.NodeSpec{PodCIDR: "10.120.1.0/24"}},
+			},
+			initialRoutes: []*cloudprovider.Route{
+				{cluster + "-01", "node-1", "10.120.0.0/24", "k8s-node-route"},
+				{cluster + "-03", "node-3", "10.120.2.0/24", "k8s-node-route"},
+			},
+			expectedRoutes: []*cloudprovider.Route{
+				{cluster + "-01", "node-1", "10.120.0.0/24", "k8s-node-route"},
+				{cluster + "-02", "node-2", "10.120.1.0/24", "k8s-node-route"},
+			},
+		},
+	}
+	for i, testCase := range testCases {
+		cloud := &fake_cloud.FakeCloud{RouteMap: make(map[string]*cloudprovider.Route)}
+		for _, route := range testCase.initialRoutes {
+			cloud.RouteMap[route.Name] = route
+		}
+		routes, ok := cloud.Routes()
+		if !ok {
+			t.Error("Error in test: fake_cloud doesn't support Routes()")
+		}
+		_, cidr, _ := net.ParseCIDR("10.120.0.0/16")
+		rc := New(routes, nil, cluster, cidr)
+		if err := rc.reconcile(testCase.nodes, testCase.initialRoutes); err != nil {
+			t.Errorf("%d. Error from rc.reconcile(): %v", i, err)
+		}
+		time.Sleep(10 * time.Millisecond)
```

Comment:
I feel gross about this. The reconciler kicks off goroutines to add/delete routes in the cloudprovider, so immediately checking the state of the fake_cloud in the main thread will show that nothing has been updated. A sleep of pretty much any length is enough to allow the background threads to update the in-memory representation, but it feels so wrong.

Any suggestions on a better way?